### PR TITLE
[RISCV] Expand mul X, C where C=2^N*(3,5,9)*(3,5,9)

### PR DESCRIPTION
**Pull request by preames**
This is a three deep expression which is deeper than we've otherwise gone for multiple expansions, but I think it's reasonable to do so. This covers mul by 50, 100, and 200 which are reasonably common naturally arising numbers.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 1]
```diff
@@ -14131,6 +14131,32 @@ static SDValue expandMul(SDNode *N, SelectionDAG &DAG,
...
+    for (uint64_t Divisor : {3, 5, 9}) {
+      if (MulAmt % Divisor != 0)
+        continue;
+      uint64_t MulAmt2 = MulAmt / Divisor;
+      // 3/5/9 * 3/5/9 * 2^N - In particular, this covers multiples
+      // of 25 which happen to be quite common.
+      for (uint64_t Divisor2 : {3, 5, 9}) {
+        if (MulAmt2 % Divisor2 != 0)
+          continue;
+        uint64_t MulAmt3 = MulAmt2 / Divisor;
```
Comment (dtcxzyw) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:14144]:
```suggestion
        uint64_t MulAmt3 = MulAmt2 / Divisor2;
```

**Review by dtcxzyw — CHANGES_REQUESTED** [review]
Miscompilation:
```
; bin/llc -mtriple=riscv64 -mattr=+m,+zba test.ll -o -
define i64 @mul108(i64 %x) nounwind {
  %mul = mul i64 %x, 108
  ret i64 %mul
}

define i64 @mul135(i64 %x) nounwind {
  %mul = mul i64 %x, 135
  ret i64 %mul
}
```
```
        .text
        .attribute      4, 16
        .attribute      5, "rv64i2p1_m2p0_zmmul1p0_zba1p0"
        .file   "test.ll"
        .globl  mul108                          # -- Begin function mul108
        .p2align        2
        .type   mul108,@function
mul108:                                 # @mul108
# %bb.0:
        sh3add  a0, a0, a0
        sh1add  a0, a0, a0
        ret
.Lfunc_end0:
        .size   mul108, .Lfunc_end0-mul108
                                        # -- End function
        .globl  mul135                          # -- Begin function mul135
        .p2align        2
        .type   mul135,@function
mul135:                                 # @mul135
# %bb.0:
        sh3add  a0, a0, a0
        sh1add  a0, a0, a0
        ret
.Lfunc_end1:
        .size   mul135, .Lfunc_end1-mul135
                                        # -- End function
        .section        ".note.GNU-stack","",@progbits
```

**Comment by preames** [comment]
> Miscompilation:

Oops, yep, you're completely correct.  Thank you.

**Review by dtcxzyw — APPROVED** [review]
LGTM. Thank you!

**Review by lukel97 — APPROVED** [review]

**Pull request merged**